From 3bddcd38b0b3accdf9ccfecd117c1fb4fc229305 Mon Sep 17 00:00:00 2001
From: progier389 <progier@redhat.com>
Date: Fri, 6 Jun 2025 15:26:52 +0200
Subject: [PATCH] Issue 6791 - crash in liblmdb during instance shutdown
 (#6793)

Sometime ns-slapd process crashes during the shutdown.
The core stacks shows that a thread attempts to use liblmdb after lmdb environment get closed
in one of thread specific data destructor callback.
The fix ensure that lmdb is not called after its environment get closed

Issue: #6791

Reviewed by: @droideck, @tbordaz (Thanks!)

(cherry picked from commit d7ba8408289601753b8f8f5298ac70ca1326b3b0)
---
 .../slapd/back-ldbm/db-mdb/mdb_instance.c     | 20 ++++++++++++++++++-
 .../slapd/back-ldbm/db-mdb/mdb_layer.h        |  1 +
 ldap/servers/slapd/back-ldbm/db-mdb/mdb_txn.c |  4 +++-
 3 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c
index 05f1e348d..b51333182 100644
--- a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c
+++ b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c
@@ -97,6 +97,9 @@ typedef struct {
 static dbmdb_dbi_t *dbi_slots;    /* The alloced slots */
 static int dbi_nbslots;           /* Number of available slots in dbi_slots */
 
+static int32_t g_mdb_env_is_open = false;
+static void dbmdb_set_is_env_open(bool is_open);
+
 /*
  * twalk_r is not available before glibc-2.30 so lets replace it by twalk
  * and a global variable (it is possible because there is a single call
@@ -720,6 +723,7 @@ int dbmdb_make_env(dbmdb_ctx_t *ctx, int readOnly, mdb_mode_t mode)
         rc =  mdb_env_open(env, ctx->home, flags, mode);
     }
     if (rc ==0) {
+        dbmdb_set_is_env_open(true);
         rc = mdb_env_info(env, &envinfo);
     }
     if (rc ==0) { /* Update the INFO file with the real size provided by the db */
@@ -752,6 +756,7 @@ int dbmdb_make_env(dbmdb_ctx_t *ctx, int readOnly, mdb_mode_t mode)
     }
     if (rc != 0 && env) {
         ctx->env = NULL;
+        dbmdb_set_is_env_open(false);
         mdb_env_close(env);
     }
     return rc;
@@ -768,6 +773,7 @@ void dbmdb_ctx_close(dbmdb_ctx_t *ctx)
          */
     }
     if (ctx->env) {
+        dbmdb_set_is_env_open(false);
         mdb_env_close(ctx->env);
         ctx->env = NULL;
     }
@@ -1735,7 +1741,7 @@ dbmdb_privdb_put(mdb_privdb_t *db, int dbi_idx, MDB_val *key, MDB_val *data)
 }
 
 
-/* Create a private database environment */
+/* Create a private database environment (used to build entryrdn during import) */
 mdb_privdb_t *
 dbmdb_privdb_create(dbmdb_ctx_t *ctx, size_t dbsize, ...)
 {
@@ -1818,3 +1824,15 @@ bail:
     }
     return db;
 }
+
+bool
+dbmdb_is_env_open()
+{
+    return (bool) slapi_atomic_load_32(&g_mdb_env_is_open, __ATOMIC_ACQUIRE);
+}
+
+static void
+dbmdb_set_is_env_open(bool is_open)
+{
+    slapi_atomic_store_32(&g_mdb_env_is_open, (int32_t)is_open, __ATOMIC_RELEASE);
+}
diff --git a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.h b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.h
index fe230d60e..fdc4a9288 100644
--- a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.h
+++ b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.h
@@ -504,6 +504,7 @@ int dbmdb_cmp_vals(MDB_val *v1, MDB_val *v2);
 dbmdb_stats_t *dbdmd_gather_stats(dbmdb_ctx_t *conf, backend *be);
 void dbmdb_free_stats(dbmdb_stats_t **stats);
 int dbmdb_reset_vlv_file(backend *be, const char *filename);
+bool dbmdb_is_env_open(void);
 
 /* mdb_txn.c */
 void shutdown_mdbtxn(void);
diff --git a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_txn.c b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_txn.c
index 74088db89..7d2dfe36e 100644
--- a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_txn.c
+++ b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_txn.c
@@ -47,7 +47,9 @@ cleanup_mdbtxn_stack(void *arg)
     slapi_ch_free((void**)&anchor);
     while (txn) {
         txn2 = txn->parent;
-        TXN_ABORT(TXN(txn));
+        if (dbmdb_is_env_open()) {
+            TXN_ABORT(TXN(txn));
+        }
         slapi_ch_free((void**)&txn);
         txn = txn2;
     }
-- 
2.49.0

